### ISSUE TITLE
"show pictures" needed again if screen rotates

### ISSUE THREAD

**Issue by dmcdivitt** [k9mail/k-9#2685]
Please search to check for an existing issue (including closed issues, for which the fix may not have yet been released) before opening a new issue: https://github.com/k9mail/k-9/issues?q=is%3Aissue

### Expected behavior
If I touch "show pictures" the state should remain if the phone is rotated (portrait to landscape or vice versa).

### Actual behavior
If the phone is rotated the "show pictures" button shows again

### Steps to reproduce
1.Display a message in the message list.
2.Touch "show pictures".
3.Rotate the phone from portrait to landscape.

### Environment
K-9 Mail version:
5.207

Android version:
5 on TV PC, 7 on ASUS Zenfone 3

Account type (IMAP, POP3, WebDAV/Exchange):
POP3

**Comment by marwendoukh** [comment]
I would like to work on this .

**Comment by dmcdivitt** [comment]
thanks, marwendoukh

**Comment by Kartik2016** [comment]
Is the issue still open? Can i work on it ?

**Comment by marwendoukh** [comment]
@Kartik2016 ,  I think this issue has been fixed by #2686

**Comment by Kartik2016** [comment]
Ok :)

**Comment by dmcdivitt** [comment]
#2686 is not the same issue. Is is a different persistence item.

**Comment by Prabashi** [comment]
As I checked for the issue, I noticed that once the orientation of the phone is changed after touching "show images", the images are showed still, but also the "show pictures" button is appeared. The problem is with appearing "show pictures" even after we using it.
Therefore the topic of the issue should be changed as ' "show changes" appears again if the screen rotates even if it was once clicked ' or something like that.
K-9 mail version: 5.206
Android version: 4.2.2

**Comment by Trogel** [comment]
`MessageContainerView.onSaveInstanceState()` should remember whether the pictures are already shown, but it seems this is never called. It would be called by `Activity.onSaveInstanceState()` (called in `MessageList.onSaveInstanceState()`), if the `MessageContainerView` object had an id, but it has none.

Adding `android:id="@+id/message_container_view"` to the root element of res/layout/message_container.xml helps: Now `MessageContainerView.onSaveInstanceState()` gets called when the screen is rotated.

Then `MessageContainerView.onRestoreInstanceState()` should restore the state after the rotation, but also this is never called. One reason seems to be that `MessageList.onRestoreInstanceState()` currently does not call `super.onRestoreInstanceState()`. But even if this is added, `MessageContainerView.onRestoreInstanceState()` does not get called.

For a test, I added `findViewById(R.id.message_container_view)` to `MessageList.onRestoreInstanceState()` and found that it returns `null`. Possibly the `MessageContainerView` object with the id `message_container_view` is not yet available in the hierarchy at that time? This could explain why its method `onRestoreInstanceState()` is not called.

Sorry, this does not yet resolve the issue, but it might help a bit.

**Comment by dmcdivitt** [comment]
Is it possible the images are shown again because they are already rendered and when changing orientation the app is merely resizing the window? Even though the screen flashes the rendering engine will not be pulling images again. I would think the initialize event is firing again unnecessarily, and because of that the "show pics" button is being recreated. Just my thoughts, thinking about the app and not looking at code.

**Comment by Trogel** [comment]
> Is it possible the images are shown again because they are already rendered and when changing orientation the app is merely resizing the window?

I don't think so. The views are not just resized, but rather recreated. At least, that is what [Handling Configuration Changes](https://developer.android.com/guide/topics/resources/runtime-changes.html) says.

**Comment by Trogel** [comment]
From the code I understand that an object of the class `MessageWebView` is used to render the message, and initially network data is blocked in that object. Therefore no images are fetched from the network, and so they cannot be rendered.

The button "show pictures" just unblocks network data and triggers a fresh rendering.

After an orientation change, network data is again blocked in the newly created `MessageWebView`. Although no images can be fetched at that time, we see that they are rendered. Apparently, fetched images are cached somewhere. (Once fetched, the images are shown even if I turn off the network and then restart k-9.)

My assumption is that showing the images after an orientation change currently works just because of some caching of fetched images. At the same time, saving/restoring the visibility state of the "show pictures" button in `MessageContainerView` and `MessageTopView` does not work as intended. So, the button is shown again.

**Comment by philipwhiuk** [comment]
Sounds likely @Trogel

**Comment by Trogel** [comment]
I added some debug output. In fact, at the time when `MessageList.onRestoreInstanceState()` is executed after an orientation change, the `MessageContainerView` object has not yet been created. This happens later in the process via some asynchronous `onLoadFinished()` callback in `MessageLoaderHelper` and `MessageTopView.showMessage()`.

But the `MessageTopView` object already exists in the hierarchy at the time of `onRestoreInstanceState()`.

I don't have the full picture yet, but would it possibly make sense to move the currently not working onSave/RestoreInstanceState for the visibility state of the "show pictures" button from `MessageContainerView` to `MessageTopView`?

**Comment by dmcdivitt** [comment]
Thanks for fixing this issue! It was a nuisance.
